### DISCUSSION TITLE
getting the "component props" in markdown

### DISCUSSION THREAD

**Discussion by Elon-Mask12** [remarkjs/remark#1253] [Q&A] [answered]
so i have this use case in my app where i wanna check if the content has a particular property (say startsWith "!!!note") - if yes, return a callout kinda component with the rest of the content.

so on creating the custom component, i get the props item, im able to convert the whole thing to text, but that loses all the links in it.
Is there anyway i can get the markdown data from the props object?

Thanks : )

**Comment by ChristianMurphy** [comment]
Welcome @Elon-Mask12! 👋 

It sounds like you are looking for https://github.com/remarkjs/remark-directive
the syntax is:

```md
:::note{attribute=value}
inner content
:::
```

and yes it can have attributes and inner content.

**Reply by Elon-Mask12** [reply]
Hey, thank you for the quick reply.
Please correct me if im wrong, but this seems more of a way to extend the markdown blocks right?
My use case isnt limited to just checking "!!!note",etc, and also i have no control over the incoming data. 

Im looking for a solution that'd let me convert the "props" item (passed to the custom components in ReactMarkdown) into markdown, so i could do the required checks and evaluate.

thanks

**Reply by ChristianMurphy** [reply]
> Please correct me if im wrong, but this seems more of a way to extend the markdown blocks right?

Right, Callouts are not a part of markdown and `!!!note` is not a part of markdown.
You need some form of extension to handle it, directives are the standard way to add features to markdown.

> My use case isnt limited to just checking "!!!note",etc, and also i have no control over the incoming data.

What is your incoming data then? It doesn't seem to be markdown.
It sounds you are dealing with a domain specific language?

> Im looking for a solution that'd let me convert the "props" item (passed to the custom components in ReactMarkdown) into markdown, so i could do the required checks and evaluate.

Plugins are how content can be shaped and props can be set in `react-markdown` and the rest of `remark` https://unifiedjs.com/learn/
It's unclear what would be "converted" the syntax you show isn't valid in any markdown flavor I've seen.
And it is unclear from your description what you even want it to do.

You can find more on how to frame your question so others can actually help you at https://github.com/remarkjs/.github/blob/main/support.md

**Reply by Elon-Mask12** [reply]
sorry about that, here's an example of what im looking for, hope that explains my requirements better.

```tsx
<ReactMarkdown
   components={{
        h1: (props: any) => HeadingRenderer(props),
      }}
>
      {markdownText}
</ReactMarkdown>
```

```tsx
export const HeadingRenderer = (props: any) => {
 const markdown =  
}
```

I'm looking for a way to get the markdown value  (only the part corresponding to the h1 tag) inside the HeadingRenderer function.

**Reply by ChristianMurphy** [reply]
This is an [XY Problem](https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem/66378#66378).

What is a "markdown value"?
And what are you trying to do with it?

Take some time to frame your question.

**Reply by Elon-Mask12** [reply]
> Take some time to frame your question.

nah im done

> What is a "markdown value"?

i literally typed that in there. 
the text in markdown format that corresponds to the h1 tag.

wdym "what are you trying to do with it", dude i just wanna check say if there is a link inside it, or whatever. just basic string operations. All i asked for is a way to convert the props argument back to markdown.

How do you fail to understand this.

**Reply by ChristianMurphy** [reply]
You are asking for help from volunteers.
Spending time helping you when you have no clue what you even want to do.
No need to be rude.

Be polite or get banned.

---

This is an [XY Problem](https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem/66378#66378).
You keep on asking about your proposed solution which isn't working instead of sharing the problem you are actually trying to solve.

You can access the markdown structure from a remark plugin as noted in my earlier reply https://github.com/remarkjs/react-markdown#architecture

If you can articulate what you want to do, and do so in a polite way, I'm happy to help more.

**Reply by Elon-Mask12** [reply]
my bad, sorrry about that. 
you guys deserve respect, and it was totally wrong by me.

so i wanna check if the h1 element has the following property:
1. starts with "starred", if yes replace the word with "recommended:" else, replace it with "basic:"
2. has a link with the domain "microsoft"
3. the number of words is a multiple of 3

based on each case i'd like to return a different component or text in the `HeadingRenderer` function

```tsx
export const HeadingRenderer = (props: any) => {
 return <h1>
  //return based on the conditions mentioned above
   </h1>
}
```

**Reply by ChristianMurphy** [reply]
Again remark plugins are your friends, it allows you to inspect and work with the markdown document in an abstracted way, which allows you to better handle edge cases (there are a lot of possible edge cases in your three examples) https://unifiedjs.com/learn/

An example of your second goal "has a link with the domain "microsoft", runnable sandbox: https://stackblitz.com/edit/github-dbm27w?file=src%2Fapp.tsx

<details>
<summary>source from sandbox</summary>

```ts
import type { Root } from 'mdast';
import Markdown from 'react-markdown';
import { visit } from 'unist-util-visit';

const markdownSource = `
# heading [microsoft link](https://microsoft.com)

# heading [google link](https://google.com)
`;

function remarkHeadingLinkPlugin() {
  return function (tree: Root) {
    // visit all the headings in the markdown document
    visit(tree, 'heading', (headingNode) => {
      let hasMicrosoftLink = false;

      // scan the contents for links
      visit(headingNode, 'link', (linkNode) => {
        const linkReference = new URL(linkNode.url);
        // match the hostname of the link
        if (linkReference.hostname === 'microsoft.com') {
          hasMicrosoftLink = true;
        }
      });

      if (hasMicrosoftLink) {
        // if no data prop exists, add it
        if (!headingNode.data) headingNode.data = {};
        // add new html property to track details
        headingNode.data.hProperties = {
          ...headingNode.data.hProperties,
          'data-has-microsoft-link': 'true',
        };
      }
    });
  };
}

function App() {
  return (
    <Markdown
      remarkPlugins={[remarkHeadingLinkPlugin]}
      components={{
        h1: ({ children, ...props }) => (
          <h1>
            {
              // do something with the attribute added in the remark plugin
              // @ts-expect-error data props are not typed
              props['data-has-microsoft-link']
                ? 'this is a microsoft heading:'
                : 'this is not a microsoft heading:'
            }{' '}
            {children}
          </h1>
        ),
      }}
    >
      {markdownSource}
    </Markdown>
  );
}
```

</details>

> starts with "starred", if yes replace the word with "recommended:" else, replace it with "basic:"
> the number of words is a multiple of 3

These two are tricky depending on what you define as "words".
Link labels and image labels can contain text but do you consider those "words"?
How do emoji, non-Latin characters, and contractions get counted or ignored as part of these?

https://github.com/syntax-tree/mdast-util-to-string and https://github.com/remarkjs/remark-retext will probably help towards this goal, but the definition of "words" will still be up to you and your stakeholders.

**Reply by Elon-Mask12** [reply, marked as answer]
Sorry about the late reply been a bit busy lately.
I did a bit of searching using the links you shared, and thanks a ton i managed to find something.
Incase anyones interested:

```tsx
import { toMdast } from "hast-util-to-mdast";
import { toMarkdown } from "mdast-util-to-markdown";

const getMarkdownFromProps = (props: any) => {
  const mdast = toMdast(props.node);
  const md = toMarkdown(mdast);
  return md;
};

```
